hetmech/degree_weight.py

```diff
@@ -368,6 +368,75 @@ def add_head_tail(metapath, indices):
     return segments
 
 
+def get_all_segments(metagraph, metapath):
+    """
+    Return all subsegments of a given metapath, including those segments that
+    appear only after early splits.
+
+    Parameters
+    ----------
+    metagraph : hetio.hetnet.MetaGraph
+    metapath : hetio.hetnet.MetaPath
+
+    Returns
+    -------
+    list
+
+    Example
+    -------
+    >>> get_all_segments(metagraph, CrCbGaDrDaG)
+    [CrC, CbG, GaDrDaG, GaD, DrD, DaG]
+    """
+    segments = get_segments(metagraph, metapath)
+    if len(segments) == 1:
+        return [metapath]
+    else:
+        all_subsegments = []
+        for segment in segments:
+            subsegments = get_all_segments(metagraph, segment)
+            next_split = subsegments if len(subsegments) > 1 else []
+            all_subsegments = all_subsegments + [segment] + next_split
+        return all_subsegments
+
+
+def order_segments(metagraph, metapaths, store_inverses=False):
+    """
+    Gives the frequencies of metapath segments that occur when computing DWPC.
+    In DWPC computation, metapaths are split a number of times for simpler computation.
+    This function finds the frequencies that segments would be used when computing
+    DWPC for all given metapaths. For the targeted caching of the most frequently
+    used segments.
+
+    Parameters
+    ----------
+    metagraph : hetio.hetnet.MetaGraph
+    metapaths : list
+        list of hetio.hetnet.MetaPath objects
+    store_inverses : bool
+        Whether or not to include both forward and backward directions of segments.
+        For example, if False: [CbG, GbC] -> [CbG, CbG], else no change.
+
+    Returns
+    -------
+    collections.Counter
+        Number of times each metapath segment appears when getting all segments.
+    """
+    all_segments = [segment for metapath in metapaths for segment in get_all_segments(metagraph, metapath)]
+    if not store_inverses:
+        # Change all instances of inverted segments to the same direction, using a first-seen ordering
+        stored_directions = set(all_segments)
+        for i, v in enumerate(all_segments):
+            if v.inverse in stored_directions:
+                if v in stored_directions:
+                    stored_directions.remove(v.inverse)
+                else:
+                    all_segments[i] = v.inverse
+    segment_counts = collections.Counter(all_segments)
+    # Keep only those which appear more than once
+    segment_counts = collections.Counter({k: v for k, v in segment_counts.items() if v > 1})
```

Comment:
Let's keep single-count segments, since I think we may want to compute metrics that would like to know this number... i.e. cacheability. However, it's a good insight that we would not want to cache these, however that exclusion can occur in the caching class.